Detectors/PHOS/workflow/src/ClusterizerSpec.cxx

```diff
@@ -0,0 +1,123 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#include "FairLogger.h"
+
+#include "DataFormatsPHOS/Digit.h"
+#include "DataFormatsPHOS/Cluster.h"
+#include "DataFormatsPHOS/PHOSBlockHeader.h"
+#include "PHOSWorkflow/ClusterizerSpec.h"
+#include "Framework/ControlService.h"
+
+using namespace o2::phos::reco_workflow;
+
+void ClusterizerSpec::init(framework::InitContext& ctx)
+{
+  LOG(DEBUG) << "[PHOSClusterizer - init] Initialize clusterizer ...";
+
+  // Initialize clusterizer and link geometry
+  mClusterizer.initialize();
+}
+
+void ClusterizerSpec::run(framework::ProcessingContext& ctx)
+{
+  if (ctx.inputs().isValid("digits")) {
+    LOG(DEBUG) << "PHOSClusterizer - run on digits called";
+
+    auto dataref = ctx.inputs().get("digits");
+    auto const* phosheader = o2::framework::DataRefUtils::getHeader<o2::phos::PHOSBlockHeader*>(dataref);
+    if (!phosheader->mHasPayload) {
+      LOG(DEBUG) << "[PHOSClusterizer - run] No more digits" << std::endl;
+      ctx.services().get<o2::framework::ControlService>().readyToQuit(framework::QuitRequest::Me);
+      return;
+    }
+
+    // auto digits = ctx.inputs().get<gsl::span<o2::phos::Digit>>("digits");
+    // results in [7968:PHOSClusterizerSpec]: [20:51:44][ERROR] Exception caught: Inconsistent serialization method for extracting span
+    auto digits = ctx.inputs().get<std::vector<o2::phos::Digit>>("digits");
+    // auto digitsTR = ctx.inputs().get<gsl::span<o2::phos::TriggerRecord>>("digitTriggerRecords");
```

Comment:
using span here instead of vector causes 
[7968:PHOSClusterizerSpec]: [20:51:44][ERROR] Exception caught: Inconsistent serialization method for extracting span

Reply:
could you send me full error message?

Reply:
here is full log: https://drive.google.com/file/d/1UYuvAhLxcGZIhs2j7pHNbeMGmwpw7XXv/view?usp=sharing
For some reason I can not attach it directly here...
    Dmitri

Reply:
looks like the TriggerRecords are not messaged via shared memory but serialized. This is strange since you've copied the class from the emcal and there it is messagable. I will need to build your branch and test myself, but will not to do this today...

Reply:
There might be problems with TR as well, but when I tried to debug a bit I found that exception is thrown while getting  Digits. With vectors (as in EMCAL) it works well.
    Thanks, Dmitri

Reply:
I've tried with the same outcomes as yours. The problem comes from the Publisher/RootTreeReader. Will try to debug tomorrow with Giulio. Meanwhile, I see that you have conflicting branchnames as PHS... and PHOS..., could you bring them to unique standard (preferaby PHS)?

Reply:
Thank you, Ruben. Concerning branch naming: I always use data producer PHS while in branch names I always use PHOS. I will cross-check.

Reply:
Hi @peressounko 
The RootTreeReader was now fixed, so you can use spans instead of vectors (but you need to add a few lines in the publisher and setting up the reader, see https://github.com/AliceO2Group/AliceO2/commit/4abc548943f4b0c871d599425fe422c895292fe8)

Cheers,
 Ruben